Jellyfin.Networking/Manager/NetworkManager.cs

```diff
@@ -0,0 +1,1283 @@
+                }
+
+                if (!IsIP4Enabled && source.AddressFamily == AddressFamily.InterNetwork)
+                {
+                    _logger.LogWarning("IPv4 is disabled in JellyFin, but enabled in the OS. This may affect how the interface is selected.");
+                }
+
+                isExternal = !IsInLocalNetwork(source);
+
+                if (MatchesPublishedServerUrl(source, isExternal, isChromeCast, out string result, out port))
+                {
+                    _logger.LogInformation("{0}: Using BindAddress {1}:{2}", source, result, port);
+                    return result;
+                }
+            }
+
+            _logger.LogDebug("GetBindInterface: Souce: {0}, External: {1}:", haveSource, isExternal);
+
+            // No preference given, so move on to bind addresses.
+            lock (_intLock)
+            {
+                if (MatchesBindInterface(source, isExternal, out string result))
+                {
+                    return result;
+                }
+
+                if (isExternal && MatchesExternalInterface(source, out result))
+                {
+                    return result;
+                }
+
+                // Get the first LAN interface address that isn't a loopback.
+                var interfaces = new NetCollection(_interfaceAddresses
+                    .Exclude(_bindExclusions)
+                    .Where(p => IsInLocalNetwork(p))
+                    .OrderBy(p => p.Tag));
+
+                if (interfaces.Count > 0)
+                {
+                    if (haveSource)
+                    {
+                        // Does the request originate in one of the interface subnets?
+                        // (For systems with multiple internal network cards, and multiple subnets)
+                        foreach (var intf in interfaces)
+                        {
+                            if (intf.Contains(source))
+                            {
+                                result = FormatIP6String(intf.Address);
+                                _logger.LogDebug("{0}: GetBindInterface: Has source, matched best internal interface on range. {1}", source, result);
+                                return result;
+                            }
+                        }
+                    }
+
+                    result = FormatIP6String(interfaces.First().Address);
+                    _logger.LogDebug("{0}: GetBindInterface: Matched first internal interface. {1}", source, result);
+                    return result;
+                }
+
+                // There isn't any others, so we'll use the loopback.
+                result = IsIP6Enabled ? "::" : "127.0.0.1";
+                _logger.LogWarning("{0}: GetBindInterface: Loopback return.", source, result);
+                return result;
+            }
+        }
+
+        /// <inheritdoc/>
+        public NetCollection GetInternalBindAddresses()
+        {
+            lock (_intLock)
```

Comment:
Is all this locking really necessary? It seems a bit extreme for reads.

Reply:
It's belt and braces since the network config settings could change at any time - either by the user, or through network events.

Reply:
But if a read locks it and uses the current interfaces to create a NetCollection and something changes immediately after, then the collection is invalid. So the locking is essentially useless?

Reply:
Was more belt and braces so things don't cause exceptions, and leave something in an unrecoverable state, rather than to ensure content remains correct.

Happy to go lock free and only lock on write.